choose testType at setup. bookshelf supports tests

abelea2 opened the pull request:
### Issues Addressed: #62 #56 

**Add testType identifier to the device setup process:**
- SetupDevice.tsx - added testType dropdown and logic to start session with testType optional
- setupDeviceSlice.tsx - setup testType initial state and type

**Bookshelf displays correct books based on testType:**
- selectBookSlice.ts --> selectAvailableBooks( ) - filters ViewableBooks array.  If pre-/post-test, filters those that don't match testType.

**Prompt logic to select correct mode for each test question:**
- readSlice.ts --> selectMode( ) - 
              - checks bookType.  If non pre-/post-test, select mode as before
              - make sure we have a question, not a narrative page.
              - tries to grab number prompt.  If it can, good.  If not, then tries to grab size prompt.
- readSlice.ts --> selectDifficulty( ) - the pre-/post-tests are not adaptive.  Therefore, if bookType is of pre-/post-test, we should keep the difficulty at "easy", since all of the Contentful pre-/post-test questions have difficulty of "easy".

abelea2 committed:
choose testType at setup. bookshelf supports tests

davidharting commented:
Initial look at your comment and the code looks great! I will take a deeper look at this tomorrow or Friday and either request changes or approve!

abelea2 commented:
### NOTE: Probably Don't "Squash and Merge" Yet!!!

**The pre-/post-test answers don't seem to save to the database correctly/at all.**

- The code in this branch seems to allow the user to see the correct books properly and answer the questions.  But we found an issue when it attempts to save the results to the firebase database.  Specifically, in ...db/index.ts --> recordAnswer( ) the AnswerDocument appears to have all of the correct information, but at _Line 59_ it doesn't seem to receive the promise from the await back and fails to add the new document to the database.

- We were hoping to get some help from you on why this could be an issue once you had the code we'd written to look at and run.

![image](https://user-images.githubusercontent.com/71240612/99578087-41cc0180-29aa-11eb-8125-d3d8f192d0ac.png)

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -103,6 +103,7 @@ export const chooseAnswerAsync = (payload: AnswerPayload): AppThunk => (
     const answer: Answer = { ...payload, mode, difficulty, pageNumber };
     dispatch(chooseAnswer(answer));
     const record = enrichAnswer(answer, state);
+    console.log(record)
```
Let's remember to remove this log before merging.

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -178,9 +179,51 @@ const selectQuestion = (state: RootState): IQuestion | null => {
+  const bookType = book?.fields.type;
+
+  //check if pre- or post-test
+  if( bookType === "pre-test" || bookType === "post-test" ){
```
Looks like you aren't using Prettier which is a code auto-formatter that this project uses.

Would you install [this extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) to your VS code? Then if you make a change (like just a space at the end of a line or whatever) and hit save, it should format the document. 

If it doesn't automatically format on save, then you can pull up the command pallete. On windows you use ctrl+shift+p or f11 to bring that up. Then type in "format document." It should appear in the list and you can hit enter to select it.

---

**Edit** I just merged https://github.com/davidharting/Adaptive-eBook/pull/67 which also gives you the option to run `yarn format` from the command line inside the top level of this project's folder and it will format all files for you.

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -178,9 +179,51 @@ const selectQuestion = (state: RootState): IQuestion | null => {
   const treatment = selectTreatment(state);
+  const book = selectBook(state)
+  const bookType = book?.fields.type;
+
```
I am a big fan of "short-circuiting" functions. The goal is to get the majority of the code indented as little as possible. So you want your "if checks" around the shorter cases to handle.

In this case, handling pre- and post-tests is the long case to handle. 

So I personally would have re-arranged the logic to do:

```ts
// Deal with the normal case first
if (bookType !== "pre-test" && bookType !== "post-test") {
  if (treatment === "mixed") {
    return state.read.randomMode;
   }
   return treatment;
}

// Now all of your pre-/post-test logic can take up the rest of the function and not be as far indented for readability. 
```

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -178,9 +179,51 @@ const selectQuestion = (state: RootState): IQuestion | null => {
+  //check if pre- or post-test
+  if( bookType === "pre-test" || bookType === "post-test" ){
+  //make sure book isn't null (makes getPage feel better)
+    if(book){
```
If `book` is null you can just return null instead of having to indent everything in inside of this.

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -178,9 +179,51 @@ const selectQuestion = (state: RootState): IQuestion | null => {
+  //make sure book isn't null (makes getPage feel better)
+    if(book){
+      //if sys.id is a question, then continue. Otherwise, it's narrative page so skip over
+      if(Book.getPage(book,selectPageNumber(state))?.sys.contentType.sys.id === "question"){
```
I think there is already a way to do this check that will be a little cleaner. 

Check out `models/BookPage.ts`. I would do:

```ts
const bookPage = Book.getPage(book, selectPageNumber(state))
const determine = BookPage.determineType(bookPage)
if (determine.type === 'page') {
  return null
}
const question = determine.question // You won't need to use `selectQuestion` if you go this route because you will aready have it from the determineType call
```

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -178,9 +179,51 @@ const selectQuestion = (state: RootState): IQuestion | null => {
+        //try to get numberPrompt, if we can't, then it's a sizePrompt
+        try{
+          if(Question.getPrompt(question,"number")){
+            return "number";
```
This try-catch approach works, but I think it can be improved. If an error gets thrown for a different reason than we thought, this logic could break. It also makes is a bit hard to follow due to the indirection.

I think we could add a function to the `models/Question.ts` to make this flow a lot easier. Maybe something like:

```ts
static getPromptType(question: IQuestion) : 'number' | 'size' | 'both' | 'none' {
  const numberPrompt = question.fields.numberPrompt
  const sizePrompt = question.fields.sizePrompt
  if (numberPrompt && sizePrompt) {
    return 'both'
  }
  if (numberPrompt) {
    return 'number'
  }
  if (sizePrompt) {
    return 'size'
  }
  return 'none'
}
```

Then, in your code here, the process should be a lot more straightforward:

```ts
const promptType = Question.getPromptType(question)
if (promptType === 'number') { return 'number' }
if (promptType === 'size' { return 'size' }
if (promptType === 'both' { throw new Error('Pre- or post-test question had two prompts instead of one.') }
if (promptType === 'none' { throw new Error('Pre- or post-test question missing prompt') }
```

davidharting commented on src/features/read/readSlice.ts:
```diff
@@ -224,7 +267,7 @@ export function selectDifficulty(state: RootState): Difficulty {
   const answers = state.read.answers;
 
-  if (!book || answers.length < NUMBER_CORRECT_TO_ADVANCE) {
+  if (!book || answers.length < NUMBER_CORRECT_TO_ADVANCE || book?.fields.type === "pre-test" || book?.fields.type === "post-test") {
```
Since you have had to do this type of check a couple of times, I think it would be worth adding a method to the book model and using it instead.

Something like:

```ts
function isAssessment(book: IBook) : boolean {
  return book.fields.type === "pre-test" || book.fields.type === "post-test"
}
```

davidharting commented on src/features/select-book/selectBookSlice.ts:
```diff
@@ -65,7 +65,18 @@ export const {
+  //define and filter array of viewable books
+  let ViewableBooks: Array<IBook> = state.content.books.filter( book => 
+    ( 
+      (book.fields.type === "pre-test" || "post-test") 
```
I think this logic actually means something slightly different from what you intended.

You unfortunately have to be pretty wordy and say `book.fields.type === "pre-test" || book.fields.type === "post-test"`. But if you do the earlier suggestion to create a `Book.isAssessment` function you could replace this bit of logic with that!

davidharting commented on src/features/select-book/selectBookSlice.ts:
```diff
@@ -65,7 +65,18 @@ export const {
 export const selectAvailableBooks = (state: RootState) => {
-  return state.content.books;
+  //define and filter array of viewable books
+  let ViewableBooks: Array<IBook> = state.content.books.filter( book => 
```
You don't really have to declare a variable here and then return it at line `79`. You could instead simply add a return statement to the front of this:


```ts
return state.content.books.filter(/* ... */)
```

and the filter call can still span multiple lines as you have now.

davidharting commented on src/features/setup-device/SetupDevice.tsx:
```diff
@@ -9,14 +9,16 @@ function NewSession() {
+    // if (childName && parentName && treatment && testType) { //empty string in Javascript is technically "falsy" therefore && testType == false
     if (childName && parentName && treatment) {
-      dispatch(startSession({ childName, parentName, treatment }));
+      dispatch(startSession({ childName, parentName, treatment, testType }));
```
Do you know if Test Type should be a require field? 

Since you didn't change line `20`, currently looks like they would be able to submit the form without specifying a test type. 

I'm not quite sure how Mary wants to roll this out. If she has test type assignments ready for everyone going forward, then maybe we should make it required? Although we would want to make sure she knows that the implication is that as soon as this is merged people could no longer get through set up without a test type.

davidharting commented on src/features/setup-device/SetupDevice.tsx:
```diff
@@ -61,6 +63,21 @@ function NewSession() {
             </Form.Control>
           </Form.Group>
+          <Form.Group>
+            <Form.Label>What type of test were you assigned?</Form.Label>
```
We may want to consult with Mary on the wording that goes here.

davidharting reviewed (commented):
Really nice work. This ended up touching quite a few different areas of the app, and you accomplished a great deal in this PR! Really, this single PR gets us most if not all the way to the goal of having assessments in the app!

I left quite a few comments with suggestions and ideas and a couple of questions as well. Sorry if you feel like I'm being picky! I can be a bit obsessed with trying to make code readable so that it is easy to understand in the future. 

I will pull down this code and take a look at the database issue you mentioned! You say they are not being saved. Are you seeing something in the chrome dev tools that made you think this, or because you can't see them show up in Metabase?

abelea2 commented:
@davidharting ,

Hi David, thanks for taking a look at the code.  We'll try to implement these revisions this week.  

Regarding the saving, we were concerned because we could not see the results showing up in Metabase.

Would it be possible to meet sometime next week to go over these revisions, once we've completed them, and to hopefully address the saving issue as well?  Let us know a day/time that would work well for you.  Thanks again!

abelea2 commented:
David,
I believe we have covered all of the above revisions in my local environment, but haven't pushed the changes yet.  We will do further testing to make sure it all works as before.

abelea2 committed:
Responded to Feedback

abelea2 committed:
Merge branch 'master' of https://github.com/davidharting/adaptive-ebook into add-support-for-pre-and-post-tests

abelea2 committed:
Updated SetupDevice Formatting

abelea2 committed:
Remove "Optional" from setupdevice screen language

davidharting committed:
For assessments, only show feedback on first two questions per section.  (#69)

Acceptance Tests
- Go through a regular book or two. Make sure feedback shows on every page.
- Go through an assessment. Make sure feedback only shows on first two pages of each section.

davidharting commented:
@abelea2 I just merged in the feedback-hiding fix we worked on today, so after the netlify build finishes you should have that feature available in the preview URL!

davidharting committed:
Do not show "feedback hidden"

I left it in there for spacing reasons - I don't want the images to jump. But Dr. Fuhs prefers that we not show any text at all when we are hiding feedback.
